compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp

```diff
@@ -17,6 +19,273 @@ namespace mlir::iree_compiler {
+    } else if (auto parallelInsertOp =
+                   dyn_cast<tensor::ParallelInsertSliceOp>(user)) {
+      expandableUsers.push_back(user);
+    } else {
+      return failure();
+    }
+  }
+  return success();
+}
+
+/// Utility to expand the pre-verified expandable users of the scf.forall
+/// output.
+static void expandVerifiedUsers(PatternRewriter &rewriter, Location loc,
+                                MLIRContext *ctx,
+                                SmallVector<Operation *> expandableUsers,
+                                SmallVector<int64_t> totalInnerSizes,
+                                SmallVector<ReassociationIndices> reIndices,
+                                scf::ForallOp forallOp) {
+  // The user expands and producer collapses need to be
+  // unflattened e.g %collapsed = tensor.collapse_shape %transposed [[0, 1], [2,
+  // 3]] : tensor<8x16x8x16xf32> into tensor<128x128xf32> can be unflattened to
+  // %collapsed = tensor.collapse_shape %transposed [[0], [1], [2], [3]] :
+  // tensor<8x16x8x16xf32> into tensor<8x16x8x16xf32> and then is consumed by
+  // the expanded parallel_insert_slice_op.
+  SmallVector<ReassociationIndices> unFlattenReassociations;
+  for (ReassociationIndices inds : reIndices) {
+    for (int64_t i : inds) {
+      unFlattenReassociations.push_back({i});
+    }
+  }
+  // compute the offsets,sizes,strides in the expanded dimensions.
+  auto computeExpandedAccess = [&](ArrayRef<OpFoldResult> mixedOffsets,
+                                   ShapedType resultType) {
+    SmallVector<OpFoldResult> expandedOffsets;
+    auto expandedOffsetsIter = expandedOffsets.begin();
+
+    for (auto [index, offset] : llvm::enumerate(mixedOffsets)) {
+      // Add zero offsets for the extra dimensions from reIndices.
+      for (size_t i = 1, e = reIndices[index].size(); i < e; ++i) {
+        expandedOffsets.push_back(getAsIndexOpFoldResult(ctx, 0));
+      }
+      // Compute the outer dimension expression.
+      AffineExpr s0, s1;
+      bindSymbols(rewriter.getContext(), s0, s1);
+      AffineExpr outerDimExpr = (s0).floorDiv(s1);
+      // Insert computed offset using affine expression.
+      expandedOffsets.insert(
+          expandedOffsetsIter,
+          affine::makeComposedFoldedAffineApply(
+              rewriter, loc, outerDimExpr,
+              {getValueOrCreateConstantIndexOp(rewriter, loc, offset),
+               rewriter.getIndexAttr(totalInnerSizes[index])}));
+
+      expandedOffsetsIter = expandedOffsets.end();
+    }
+    ArrayRef<int64_t> expandedShape = resultType.getShape();
+    SmallVector<OpFoldResult> expandedSizes;
+    for (auto size : expandedShape) {
+      expandedSizes.push_back(getAsIndexOpFoldResult(ctx, size));
+    }
+    SmallVector<OpFoldResult> expandedStrides(resultType.getRank(),
+                                              rewriter.getIndexAttr(1));
+    return std::make_tuple(expandedOffsets, expandedSizes, expandedStrides);
```

Comment:
```suggestion
    return {expandedOffsets, expandedSizes, expandedStrides};
```

Reply:
I tried this one first, it threw a compiler error about not being able to infer the type.

Reply:
I'd add casts in this case then, since otherwise you have some implicit conversion over the tuple types instead which seem more scary

Reply:
I got it working by being explicit about the return type of the lambda